pypesto/hierarchical/problem.py

```diff
@@ -0,0 +1,437 @@
+    }
+
+    for par in inner_parameters:
+        if par.inner_parameter_type not in [
+            InnerParameterType.SCALING,
+            InnerParameterType.OFFSET,
+        ]:
+            continue
+        if par.inner_parameter_id in coupled_pars:
+            par.coupled = True
+
+    return AmiciInnerProblem(xs=inner_parameters, data=data, edatas=edatas)
+
+
+def inner_parameters_from_parameter_df(
+    df: pd.DataFrame,
+) -> List[InnerParameter]:
+    """
+    Create list of inner free parameters from PEtab parameter table.
+
+    Inner parameters are those that have a non-empty `parameterType` in the
+    PEtab problem.
+    """
+    # create list of hierarchical parameters
+    df = df.reset_index()
+
+    for col in (PARAMETER_TYPE,):
+        if col not in df:
+            df[col] = None
+
+    parameters = []
+
+    for _, row in df.iterrows():
+        if not row[ESTIMATE]:
+            continue
+        if petab.is_empty(row[PARAMETER_TYPE]):
+            continue
+        parameters.append(
+            InnerParameter(
+                inner_parameter_id=row[PARAMETER_ID],
+                inner_parameter_type=row[PARAMETER_TYPE],
+                scale=row[PARAMETER_SCALE],
+                lb=row[LOWER_BOUND],
+                ub=row[UPPER_BOUND],
+            )
+        )
+
+    return parameters
+
+
+def ixs_for_measurement_specific_parameters(
+    petab_problem: 'petab.Problem',
+    amici_model: 'amici.Model',
+    x_ids: List[str],
+) -> Dict[str, List[Tuple[int, int, int]]]:
+    """
+    Create mapping of parameters to measurements.
+
+    Returns
+    -------
+    A dictionary mapping parameter ID to a List of condition, time, observable
+    index tuples in which this output parameter is used.
+    """
+    ixs_for_par = {}
+    observable_ids = amici_model.getObservableIds()
+
+    simulation_conditions = (
+        petab_problem.get_simulation_conditions_from_measurement_df()
+    )
+    for condition_ix, condition in simulation_conditions.iterrows():
+        df_for_condition = petab.get_rows_for_condition(
+            measurement_df=petab_problem.measurement_df, condition=condition
+        )
+
+        timepoints = sorted(df_for_condition[TIME].unique().astype(float))
+        timepoints_w_reps = _get_timepoints_with_replicates(
+            measurement_df=df_for_condition
+        )
+
+        for time in timepoints:
+            # subselect for time
+            df_for_time = df_for_condition[df_for_condition[TIME] == time]
+            time_ix_0 = timepoints_w_reps.index(time)
+
+            # remember used time indices for each observable
+            time_ix_for_obs_ix = {}
+
+            # iterate over measurements
+            for _, measurement in df_for_time.iterrows():
+                # extract observable index
+                observable_ix = observable_ids.index(
+                    measurement[OBSERVABLE_ID]
+                )
+
+                # update time index for observable
```

Comment:
I feel like this piece of code warrants a bit more explanation.

Reply:
tried to improve